src/systems/filecoin_vm/runtime/runtime.go

```diff
@@ -1,81 +1,424 @@
+	InvokeMethod(rt Runtime, method actor.MethodNum, params actor.MethodParams)
 }
 
-func ErrorInvocOutput(tree st.StateTree, ec msg.ExitCode) InvocOutput {
-	return InvocOutput{
-		OutTree:     tree,
-		GasUsed:     gascost.CodeLookupFail,
-		ExitCode:    exitcode.InvalidMethod,
-		ReturnValue: nil,
+const (
+	Reserved_NoopMethod        actor.MethodNum = 0
+	Reserved_CronMethod        actor.MethodNum = 1
+	Reserved_ConstructorMethod actor.MethodNum = 2
+)
+
+func Runtime_Make(
+	globalState st.StateTree,
+	actorAddress addr.Address,
+	valueSupplied actor.TokenAmount,
+	gasRemaining msg.GasAmount) Runtime {
+
+	actorStateInit := globalState.GetActor(actorAddress).State()
+
+	return &Runtime_I{
+		_globalStateInit_:        globalState,
+		_globalStatePending_:     globalState,
+		_running_:                false,
+		_actorAddress_:           actorAddress,
+		_actorStateAcquired_:     false,
+		_actorStateAcquiredInit_: actorStateInit.State(),
+
+		_valueSupplied_:    valueSupplied,
+		_gasRemaining_:     gasRemaining,
+		_numValidateCalls_: 0,
+		_numReturnCalls_:   0,
+		_messageQueue_:     []MessageQueueItem{},
+		_output_:           nil,
 	}
 }
 
-func (r *Runtime_I) Fatal(string) Runtime_Fatal_FunRet {
+func _generateActorAddress(creator addr.Address, nonce actor.CallSeqNum) addr.Address {
+	// _generateActorAddress computes the address of the contract,
+	// based on the creator (invoking address) and nonce given.
+	// TODO: why is this needed? -- InitActor
+	// TODO: this has to be the origin call. and it's broken: could yield the same address
+	//       need a truly unique way to assign an address.
 	panic("TODO")
 }
 
-func (r *Runtime_I) Send(to addr.Address, method actor.MethodNum, params actor.MethodParams, value actor.TokenAmount) msg.MessageReceipt {
+func (rt *Runtime_I) CreateActor(codeCID actor.CodeCID, constructorParams actor.MethodParams) Runtime_CreateActor_FunRet {
+	rt.ValidateCallerIs(InitActorAddr)
+	// TODO: _generateActorAddress
+	// TODO: finish
 	panic("TODO")
 }
 
-func (s *VMState_I) Epoch() st.Epoch {
-	panic("TODO")
+func (h *ActorStateHandle_I) UpdateRelease(newStateCID ActorSubstateCID) {
+	h._rt().Impl()._updateReleaseActorState(newStateCID)
+}
+
+func (h *ActorStateHandle_I) Release(checkStateCID ActorSubstateCID) {
+	h._rt().Impl()._releaseActorState(checkStateCID)
+}
+
+func (h *ActorStateHandle_I) Take() ActorSubstateCID {
+	if h._initValue().Which() == ActorStateHandle__initValue_Case_None {
+		h._rt().Impl()._apiError("Must call Take() only once on actor substate object")
+	}
+	ret := ActorSubstateCID(h._initValue().As_Some())
+	h._initValue_ = ActorStateHandle__initValue_Make_None(&ActorStateHandle__initValue_None_I{})
+	return ret
+}
+
+func (rt *Runtime_I) _updateReleaseActorState(newStateCID ActorSubstateCID) {
+	rt._checkRunning()
+	rt._checkActorStateAcquired()
+
+	rt._globalStatePending_ = rt._globalStatePending().WithActorState(rt._actorAddress(), newStateCID)
+	rt._actorStateAcquired_ = false
```

Comment:
All this state handling shouldn't be in the runtime object itself. likely a subobject (`VMState`?), or the invocation

Reply:
The runtime has a lot of things and doing this here makes it harder to reason about.

Reply:
(moved state into separate VMContext struct as per discussion)